File: ios/Faraday/Faraday/Crypto/Seal.swift
import CryptoKit
import Foundation

enum SealedEnvelope {
    static func seal(to recipientIK: Data, aad: Data, plaintext: Data) throws -> Data {
        let eph = Curve25519.KeyAgreement.PrivateKey()
        let pub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: recipientIK)
        let zz = try eph.sharedSecretFromKeyAgreement(with: pub).data
        let enc = eph.publicKey.rawRepresentation
        let shared = try kemShared(zz: zz, enc: enc, pkR: recipientIK)
        let keys = try sealKeys(shared: shared, aad: aad)
        let ct = try FaradayAEAD.seal(key: keys.key, nonce: keys.nonce, aad: aad, plaintext: plaintext)
        var out = Data([FaradayKDF.sealVersion])
        out.append(enc)
        out.append(ct)
        return out
    }

    static func open(recipientIK: Curve25519.KeyAgreement.PrivateKey, aad: Data, blob: Data) throws -> Data {
        guard blob.count >= 1 + 32 + 16, blob[0] == FaradayKDF.sealVersion else {
            throw FaradayCryptoError.seal
        }
        let enc = blob.subdata(in: 1..<33)
        let ct = blob.suffix(from: 33)
        let ephPub = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: enc)
        let zz = try recipientIK.sharedSecretFromKeyAgreement(with: ephPub).data
        let shared = try kemShared(zz: zz, enc: enc, pkR: recipientIK.publicKey.rawRepresentation)
        let keys = try sealKeys(shared: shared, aad: aad)
        return try FaradayAEAD.open(key: keys.key, nonce: keys.nonce, aad: aad, ciphertext: Data(ct))
    }

    private static func kemShared(zz: Data, enc: Data, pkR: Data) throws -> Data {
        var info = Data(FaradayKDF.infoDHKEM.utf8)
        info.append(enc)
        info.append(pkR)
        return try FaradayKDF.hkdf(ikm: zz, salt: FaradayKDF.zeros32, info: info, length: 32)
    }

    private static func sealKeys(shared: Data, aad: Data) throws -> (key: Data, nonce: Data) {
        var info = Data(FaradayKDF.infoSeal.utf8)
        info.append(aad)
        let okm = try FaradayKDF.hkdf(ikm: shared, salt: FaradayKDF.zeros32, info: info, length: 44)
        return (okm.prefix(32), okm.suffix(12))
    }
}

File: server/internal/crypto/seal.go
package crypto

import (
	"crypto/ecdh"
	"crypto/rand"
	"fmt"
)

// SealTo wraps plaintext in a single-shot anonymous DH envelope for
// recipientIK (X25519 public). This is DHKEM(X25519, HKDF-SHA256) + AES-256-GCM
// in the style of RFC 9180 HPKE Base mode. The relay sees only dest mailbox +
// this blob — not the sender identity, not the inner Signal headers.
func SealTo(recipientIK, aad, plaintext []byte) ([]byte, error) {
	eph, err := ecdh.X25519().GenerateKey(rand.Reader)
	if err != nil {
		return nil, err
	}
	pub, err := ecdh.X25519().NewPublicKey(recipientIK)
	if err != nil {
		return nil, err
	}
	zz, err := eph.ECDH(pub)
	if err != nil {
		return nil, err
	}
	enc := eph.PublicKey().Bytes()
	shared, err := kemShared(zz, enc, recipientIK)
	if err != nil {
		return nil, err
	}
	key, nonce, err := sealKeys(shared, aad)
	if err != nil {
		return nil, err
	}
	ct, err := sealAESGCM(key, nonce, aad, plaintext)
	if err != nil {
		return nil, err
	}
	out := make([]byte, 0, 1+32+len(ct))
	out = append(out, SealVersion)
	out = append(out, enc...)
	out = append(out, ct...)
	return out, nil
}

func OpenSealed(recipientIKPriv *ecdh.PrivateKey, aad, blob []byte) ([]byte, error) {
	if len(blob) < 1+32+16 {
		return nil, fmt.Errorf("seal: short")
	}
	if blob[0] != SealVersion {
		return nil, fmt.Errorf("seal: unsupported version")
	}
	enc := blob[1:33]
	ct := blob[33:]
	ephPub, err := ecdh.X25519().NewPublicKey(enc)
	if err != nil {
		return nil, err
	}
	zz, err := recipientIKPriv.ECDH(ephPub)
	if err != nil {
		return nil, err
	}
	shared, err := kemShared(zz, enc, recipientIKPriv.PublicKey().Bytes())
	if err != nil {
		return nil, err
	}
	key, nonce, err := sealKeys(shared, aad)
	if err != nil {
		return nil, err
	}
	return openAESGCM(key, nonce, aad, ct)
}

func kemShared(zz, enc, pkR []byte) ([]byte, error) {
	// RFC 9180 DHKEM ExtractAndExpand: IKM = DH, kem_context = enc || pkR
	ctx := append(append([]byte(nil), enc...), pkR...)
	return hkdfSHA256(zz, zeros32, append([]byte(InfoDHKEM), ctx...), 32)
}

func sealKeys(shared, aad []byte) (key, nonce []byte, err error) {
	info := append([]byte(InfoSeal), aad...)
	okm, err := hkdfSHA256(shared, zeros32, info, 44)
	if err != nil {
		return nil, nil, err
	}
	return okm[:32], okm[32:], nil
}
